### ISSUE TITLE
always providing acceleration signatures

### ISSUE THREAD

**Issue by psychocoderHPC** [alpaka-group/alpaka#180]
Currently alpaka disable all accelerator signatures if a accelerator is disabled in the cmake file.
This is a problem if the user would write specialization in there code e.g. for a kernel or a trait.
Currently the only way is to use `#ifdef` in the user code to disable all specializations of accelerators.
One goal of alpaka is to avoid code control by `#ifdef` in the user code because than it is hard to maintain.

This [example](https://github.com/BenjaminW3/matmul/blob/develop/include/matmul/par/AlpakaTiling.hpp#L44-L98) shows traits which are specialized for each accelerator. 
If alpaka always provide the signatures but without implementation of each accelerator the `#ifdef` code control can removed.

**Comment by BenjaminW3** [comment]
> One goal of alpaka is to avoid code control by #ifdef in the user code because than it is hard to maintain.

No, I do not think that this exactly is the goal or that it is possible at all. The goal was to move those `ifdef`s to the specializations that are used in all kernels, instead of having them around each and every kernel itself.

Your idea will only work in your special case and will not be possible in general.
Most specializations will rely on accelerator details which will not be available if 
* the compiler does not support them (e.g CUDA built-in variables, OpenMP pragmas/functions, ...)
* the accelerator members are not available
* the libraries are not available on the PC (CUDA, Boost.fiber, ...)

For such OpenMP, CUDA and similar accelerator specializations there will (nearly) always be a need for `#ifdef`s.

**Comment by psychocoderHPC** [comment]
I will reopen this issue, based on what I wrote https://github.com/ComputationalRadiationPhysics/alpaka/issues/175#issuecomment-413812625 we need to become more user friendly and provide ways to avoid `#ifdefs` in the code.
How a possible solution looks like can be discussed here.

CC-ing: tdd11235813

**Comment by psychocoderHPC** [comment]
CC-ing: @tdd11235813

**Comment by tdd11235813** [comment]
@psychocoderHPC can you give a small example?
At the moment I only have the following in mind. If you want to define, just for an example, a back-end specific trait for size type, just as an example:

One way is to use the macros. Nobody would like this that much.
```C++
struct SizeType {
#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED && BOOST_LANG_CUDA
  using type = std::uint32_t;
#elif defined(ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED) && !BOOST_LANG_CUDA
  using type = size_t;
// # ...
#endif
};
```

If I see it correctly, the current back-end types rely on two template arguments and are only enabled, if it is enabled by cmake configuration.
```C++
struct SizeType<alpaka::acc::AccGpuCudaRt<alpaka::dim::DimInt<1u>, size_t> > {
  using type = size_t;
};
```

So, what you probably suggest, is to provide meta types for the back-ends?:
```C++
template<typename AccType> 
struct SizeType;

struct SizeType<alpaka::AccCudaType> {
  using type = size_t;
};
// ...
using SizeType_t = typename SizeType< alpaka::AccCudaType >::type;
```
?

**Comment by bernhardmgruber** [comment]
Is this solved by the accelerator tags now? #1804

**Comment by SimeonEhrig** [comment]
Yes, it solves the issue.
